File: java/ProductOfArrayExceptSelf238.java
package java;

import java.util.Arrays;

/*
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].
The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
You must write an algorithm that runs in O(n) time and without using the division operation.
*/
public class ProductOfArrayExceptSelf238 {
    public static void main(String[] args) {
        System.out.println(Arrays.toString(productExceptSelf(new int[]{1, 2, 3, 4})));
        System.out.println(Arrays.toString(productExceptSelf(new int[]{-1, 1, 0, -3, 3})));
    }

    public static int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];
        int product = 1;
        for (int i = 0; i < nums.length; i++) {
            result[i] = product;
            product *= nums[i];
        }

        product = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            result[i] *= product;
            product *= nums[i];
        }

        return result;

//        int[] result = new int[nums.length];
//        for (int i = 0; i < nums.length; i++) {
//            int product = 1;
//            for (int j = 0; j < nums.length; j++) {
//                if (i != j) {
//                    product *= nums[j];
//                }
//            }
//            result[i] = product;
//        }
//        return result;
    }
}

File: python/ProductOfArrayExceptSelf238.py
# Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements
# of nums except nums[i]. The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer. You
# must write an algorithm that runs in O(n) time and without using the division operation
from typing import List


def productExceptSelf(nums: List[int]) -> List[int]:
    result, product = [], 1
    for num in nums:
        result.append(product)
        product *= num
    product = 1
    for i in range(len(nums) - 1, -1, -1):
        result[i] *= product
        product *= nums[i]
    return result


if __name__ == "__main__":
    print(productExceptSelf([1, 2, 3, 4]))
    print(productExceptSelf([-1, 1, 0, -3, 3]))
    print(productExceptSelf([1, 2, 3, 4, 5]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]))
    print(productExceptSelf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]))
